RavenDB-23792 - Allow to query by a stored vector by specifying document id

ayende opened the pull request:
- Allow to query vector searching using a document id

from 'Items' where vector.search(Vector, embedding.forDoc($p0))

### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-23792 


### Type of change

- [x] New feature

### How risky is the change?

- [x] Low 

### Backward compatibility

- [x] Non breaking change

### Is it platform specific issue?

- [x] No

### Documentation update

- [x] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works

### Testing by RavenDB QA team

- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [x] No

### UI work

- [x] No UI work is needed

ayende committed:
RavenDB-23792
- Allow to query vector searching using a document id

from 'Items' where vector.search(Vector, embedding.for($p0))

arekpalinski commented on src/Raven.Client/Documents/Queries/Vector/VectorFieldFactory.cs:
```diff
@@ -281,20 +286,25 @@ public void ByEmbedding<T>(RavenVector<T> embedding) where T : unmanaged
+    /// <summary>
+    /// Use an existing vector belonging to the specified document
+    /// </summary>
+    public void ForDocument(string documentId);
```
Please check the warning here

arekpalinski commented on src/Raven.Client/Constants.cs:
```diff
@@ -597,6 +597,7 @@ internal static class VectorSearch
         {
             private const string EmbeddingPrefix = "embedding.";
 
+            internal const string EmbeddingForDocument = EmbeddingPrefix + "for";
```
`embedding.for()` doesn't have any indication that it's going to accept a document ID as parameter. Maybe `embedding.forDoc()` ?

ayende replied:
Yes, agreed

arekpalinski commented:
Please check the following test failures:
`SlowTests.Server.Documents.AI.Embeddings.LoadVectorTests.CanIndexSingleVectorGeneratedByEtl`
`SlowTests.Server.Documents.AI.Embeddings.LoadVectorTests.CanIndexSingleVectorGeneratedByEtlJs`
 
 other two tests are already investigated by @Lwiel

ayende committed:
RavenDB-23792
- Fixing embedding generation - skip whitespace / empty strings
- Fixing text chunking for very large texts - since there is a high cost of that

Lwiel commented on src/Raven.Server/Documents/AI/Embeddings/EmbeddingsGenerator.cs:
```diff
@@ -131,6 +131,9 @@ public PutDocumentEmbeddings GenerateEmbeddingAsync(DocumentsOperationContext do
                      List<ReadOnlyMemory<byte>> cachedEmbeddingsBuffers = [];
                      foreach (var chunkedValue in TextChunker.Chunk(value, chunking))
                      {
+                         if(string.IsNullOrWhiteSpace(chunkedValue))
```
Do we want to always skip? Some providers support it. What if someone already has an index with vector corresponding to empty string?

ayende replied:
The problem is that empty or null string is not something that you really can compare _to_.
How close is `""` to `1` or `'b` ? 
I think it is rare enough that we shouldn't try - because otherwise for each provider you'll need to map that behavior, etc.

Lwiel replied:
It handles [this ticket](https://issues.hibernatingrhinos.com/issue/RavenDB-23946/Handle-Empty-Strings-in-Embeddings-Generation)

ayende committed:
RavenDB-23792
- PR comments - using embedding.forDoc()
- Fixing warning
- Ensure that we only check if we have vectors _after_ we checked if we have a task defined

arekpalinski commented:
Studio tests are failing CC @kalczur

kalczur commented:
@arekpalinski It's in `where.spec.ts`. I will check it

maciejaszyk commented:
FIY: This PR requires some grammar adjustments.
cc @arekpalinski @kalczur

maciejaszyk committed:
RavenDB-23792 Studio autocomplete - support for `embedding.forDoc()`

maciejaszyk commented:
@arekpalinski I've fixed the studio and the grammar.
